ARMmbed/mbedtls
# PKCS7 Generation - RFC 2315

Hello everyone,

This pull request is for the generation of a PKCS7. It is dependent on [Nayna's PKCS7 Parser PR](https://github.com/ARMmbed/mbedtls/pull/3431) .  Until her work gets merged, I figured I would keep our header and source files separate (I am also rather new to this sort of thing and I don't want to mess up anything she is working on). 
I appreciate any and all feedback, I assume there will be lots of things wrong so I welcome all criticism. 

Thanks for your patience and have a great holiday!
Nick

## Description
This is a new feature that allows for the generation of a PKCS7 as described by [RFC  2315](https://tools.ietf.org/html/rfc2315) . A PKCS7 can contain many different types of data. This PR allows for just content of type SignedData. The new function `mbedtls_pkcs7_create` takes in a data buffer, RSA key pairs, and a message digest type and returns a PKCS7. This PKCS7 is able to be parsed by `pkcs7_ parse_der`. The `mbedtls_pkcs7_create` function is able to receive a list of public and private keys to allow the message digest to be signed with several signers. But, please note, at the moment `pkcs7_parse_der` only accepts single signers. I additionally created a function `mbedtls_convert_pem_to_der` , this function is used by my test cases and is not directly relevant to any work done by the PKCS7 generator. Any help with finding an already available function would be greatly appreciated ( I am sure `mbedtls_convert_pem_to_der` does not need to be included in the final PR).


## Status
**READY**

## Requires Backporting
Nope

## Migrations
None. This is a new feature.

## Additional comments
I understand that my function for PKCS7 generation is a bit stack heavy from all the nested function calls. I hope this is okay. I figured it maintained clarity while also adhering to mbedtls's way of writing backwards. 

Additionally, I did not add any new error codes, I felt Nayna added all the PKCS7 errors that I would need. Please, let me know if this is unacceptable. I also reused her data files for my tests.

I am sure that I have lots to work on until this gets merged. I apologize if this is not perfect yet. I just figured I might as well get some feedback so I do not spend too much time on things that do not meet mbedtls's coding standards. 


## Steps to test or reproduce
I added a few basic test cases as a test suite, I am sure I will need to add more later on. I just wanted to make sure the general procedure was correct. Run with:
`cd tests && ./test_suite_pkcs7_write`
